text/0000-date-api.md

```diff
@@ -0,0 +1,83 @@
+- Feature Name: date_api
+- Start Date: 2018-03-21
+- RFC PR: 
+- Neon Issue: 
+
+# Summary
+[summary]: #summary
+
+This proposal defines an API for creating and inspecting JavaScript `Date` objects.
+
+# Motivation
+[motivation]: #motivation
+
+The `Date` type is part of the standard JavaScript library and should be exposed to Rust.
+
+Example:
+
+```rust
+let date: Handle<JsDate> = JsDate::new(1431673200000)?;
+let value: f64 = date.value(); // 1431673200000.0
+```
+
+# Pedagogy
+[pedagogy]: #pedagogy
+
+This is a relatively straightforward API and can be explained with the API docs.
+
+# Details
+[details]: #details
+
+## `JsDate`
+
+This API adds a `JsDate` type to `neon::types`:
+
+```rust
+impl JsDate {
+    pub const MIN_VALID_VALUE: f64 = -8640000000000000;
+    pub const MAX_VALID_VALUE: f64 = 8640000000000000;
+
+    pub fn new<'a, C: Context<'a>, V: Into<f64>>(_: &mut S, value: V) -> JsResult<JsDate>;
```

Comment:
I don't think this should be a `JsResult` since the `Date` API shouldn't throw unless we are already in a throwing state.

I do think it should return an `Error` if the value is out of bounds. I don't think it's safe to depend on being able to get the original value back out since that's not part of the contract/spec.

```rust
impl JsDate {
    /// Produces a _valid_ `Date` or a `DateError`
    pub fn new<'a, C: Context<'a>, V: Into<f64>>(_: &mut S, value: V) -> Result<Handle<'a, JsDate>, DateError>;

    /// Produces a `Date` which _could_ be invalid
    pub fn new_unchecked<'a, C: Context<'a>, V: Into<f64>>(_: &mut S, value: V) -> Handle<'a, JsDate>;
}

enum DateError {
    OutOfRangeLow,
    OutOfRangeHigh,
}

impl Error for DateError {}

impl<'a> JsResultExt<'a, JsString> for Result<Handle<'a, JsDate>, DateError> {
    /// Creates a `RangeError` on error
    fn or_throw<'b, C: Context<'b>>(self, cx: &mut C) -> JsResult<'a, JsDate>;
}
```